### DISCUSSION TITLE
How to install pre-requisites for templates?

### DISCUSSION THREAD

**Discussion by felipecrs** [twpayne/chezmoi#1817] [Q&A]
Some of my templates requires `git` and `curl`. As chezmoi executes the templates (computing source state) before executing `run_before`, what is the recommend way of installing the tools I need in the templates?

This is somewhat related to https://github.com/twpayne/chezmoi/issues/1616, despite this applies to every template, not only to `.chezmoiexternal.yaml`.

**Comment by twpayne** [comment]
If your templates require `git` and `curl` then they are no longer templates.

Some suggestions:
* Wrap your call to `chezmoi` in a script that installs `git` and `curl` before calling chezmoi.
* Use a `run_once_before_00_` script to install `git` and `curl` before your other `run_once_before_` scripts get executed (chezmoi executes templates lazily when possible).
* If your needs extend beyond managing your personal dotfiles, then consider switching to a whole system management tool like Ansible, Chef, Puppet, or Salt, as [described in the FAQ](https://github.com/twpayne/chezmoi/blob/master/docs/FAQ.md#can-i-use-chezmoi-to-manage-files-outside-my-home-directory).

**Reply by felipecrs** [reply]
Regarding your second suggestion, for some reason I can't make it work. I have a script to install the pre-requisites very early:

- `.chezmoiscripts/run_before_20-install-prerequisites.sh.tmpl`: https://github.com/felipecrs/dotfiles/blob/b345e7d3b463cd2e9a6321109d4cbfdf8faa6905/home/.chezmoiscripts/run_before_20-install-prerequisites.sh.tmpl

It however runs after my other one named `run_before_10-migrate.sh.tmpl`, and it's not a `run_once_before_`, it's a `run_before_`. Could it be due to some of these things?

If needed, can be reproduced with:

```console
$ docker run --rm -it --init --user vscode --env CODESPACES=true mcr.microsoft.com/vscode/devcontainers/base:ubuntu-20.04 bash

$ sudo apt purge git -y && cd && sh -c "$(wget -qO- git.io/chezmoi)" -- init felipecrs --apply
chezmoi: /home/vscode/.local/share/chezmoi/home/.chezmoiexternal.yaml: template: get-github-head-revision:7:17: executing "get-github-head-revision" at <output "git" "ls-remote" (printf "https://github.com/%s.git" $repo) "HEAD">: error calling output: exec: "git": executable file not found in $PATH
```

**Reply by mystygage** [reply]
Did you find a solution for this? I tried something similar for [macOS in my dotfiles](https://github.com/mystygage/dotfiles/blob/main/.chezmoiscripts/run_once_before_00-macos-init.sh.tmpl#L40) and the only reason the script `run_once_before_00-macos-init.sh.tmpl` is a template is to include the check `if ne .chezmoi.os "darwin"`.

**Reply by felipecrs** [reply]
You can put the template call in the .chezmoiignore file instead and keep the init.sh script without being a template, I think.

**Reply by twpayne** [reply]
Since chezmoi v2.25.0, chezmoi also sets environment variables like `CHEZMOI_OS` when running scripts, so you can replace your use of a template with shell conditionals, e.g.

```bash
if [ "${CHEZMOI_OS}" == "darwin" ]; then
    # ...
fi
```

**Reply by felipecrs** [reply]
I think this whole thing deserves a FAQ entry, something like:

## How to install pre-requisites for templates?

Use a `run_before` script that is **not** a template. Something like:

```bash
#!/bin/bash

set -eu

# Install curl if it's not already installed
if ! command -v curl >/dev/null; then
  sudo apt update
  sudo apt install -y curl
fi
```

Chezmoi will make sure to execute it before templating other files. You can use environment variables to inject data into your scripts and avoid using templates therefore.

**Reply by twpayne** [reply]
Very good idea @felipecrs. Could you open a PR to add the FAQ entry?

**Reply by felipecrs** [reply]
Sure, let on me.

**Reply by felipecrs** [reply]
@twpayne I'm only trying to integrate this in my dotfiles today, and it doesn't seem to work as documented (https://github.com/felipecrs/dotfiles/pull/113).

Nevertheless, we can now define pre-read-source-state hooks for this purpose.

So, I would recommend updating the documentation to point to the `pre-read-source-state` hook approach instead.

**Reply by twpayne** [reply]
I can't tell from https://github.com/felipecrs/dotfiles/pull/113 what you're trying to achieve, what you expect to happen, or what is actually happening.

Could you make a minimal example that demonstrates the issue or propose an exact documentation change?

**Reply by felipecrs** [reply]
I was trying to achieve what this discussion concluded in, back in 2022: https://www.chezmoi.io/user-guide/frequently-asked-questions/usage/#how-do-i-install-pre-requisites-for-templates

The documentation says chezmoi will run the `run_before` script before templating other files as long the script is not a template itself. But it doesn't seem to be the case, at least for such minimal reproduction repo I built.

Here is the minimal repository: https://github.com/felipecrs/chezmoi-repro-template-pre-requisites

```console
# enter in some container to avoid polluting your environment
❯ docker run --rm -it mcr.microsoft.com/devcontainers/base:ubuntu-22.04

$ sudo apt remove curl -y

$ git clone https://github.com/felipecrs/chezmoi-repro-template-pre-requisites

$ chezmoi-repro-template-pre-requisites/install.sh
Installing chezmoi to '/root/.local/bin/chezmoi'
info found version 2.48.0 for latest/linux/amd64
info installed /root/.local/bin/chezmoi
Running 'chezmoi init --apply --source=/home/felipecrs/repos/chezmoi-repro-template-pre-requisites'
💡 Tip: you can re-enter your name and email with `chezmoi init --data=false`.
chezmoi: /home/felipecrs/repos/chezmoi-repro-template-pre-requisites/.chezmoiexternal.yaml.tmpl: template: /home/felipecrs/repos/chezmoi-repro-template-pre-requisites/.chezmoiexternal.yaml.tmpl:1:10: executing "/home/felipecrs/repos/chezmoi-repro-template-pre-requisites/.chezmoiexternal.yaml.tmpl" at <output "curl" "--version">: error calling output: curl --version: exec: "curl": executable file not found in $PATH
```
